src/main/java/org/terasology/dynamicCities/region/ResourceProvider.java

```diff
@@ -1,51 +1,36 @@
-/*
- * Copyright 2016 MovingBlocks
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
 package org.terasology.dynamicCities.region;
 
 import org.terasology.dynamicCities.facets.ResourceFacet;
 import org.terasology.dynamicCities.facets.RoughnessFacet;
 import org.terasology.engine.world.generation.Border3D;
-import org.terasology.engine.world.generation.FacetProvider;
 import org.terasology.engine.world.generation.GeneratingRegion;
 import org.terasology.engine.world.generation.Produces;
+import org.terasology.engine.world.generation.ScalableFacetProvider;
 import org.terasology.engine.world.generator.plugin.RegisterPlugin;
+import org.terasology.math.TeraMath;
 
-@RegisterPlugin
-@Produces(ResourceFacet.class)
 /**
- * This facet will be used to store information about the resources in its region
- * The actual data will be set through CompatibleRasterizer
- * It will be saved as a Component in the RegionEntity
+ * This facet will be used to store information about the resources in its region.
+ *
+ * The actual data will be set through CompatibleRasterizer. It will be saved as a Component in the RegionEntity.
  */
-public class ResourceProvider implements FacetProvider {
+@RegisterPlugin
+@Produces(ResourceFacet.class)
+public class ResourceProvider implements ScalableFacetProvider {
 
     private final int gridSize = 32;
 
-
     @Override
-    public void setSeed(long seed) { }
+    public void setSeed(long seed) {
+    }
 
     @Override
-    public void process(GeneratingRegion region) {
-
+    public void process(GeneratingRegion region, float scale) {
         Border3D border = region.getBorderForFacet(RoughnessFacet.class);
-        ResourceFacet facet = new ResourceFacet(region.getRegion(), border, gridSize);
+        ResourceFacet facet = new ResourceFacet(region.getRegion(), border, TeraMath.ceilToInt(gridSize / scale));
```

Comment:
Is division correct here, or should it be a multiplication? :zany_face:

Reply:
If the result of the `ResourceProvider` only has an effect in the actual "real" chunks, maybe we should bail out early if the `scale` is not 1?

Reply:
Well, we need *a* `ResourceFacet` so that `CompatibleRasterizer`s can add resources to it. I guess an alternative would be to make `CompatibleRasterizer` check for a null resource facet and ignore it.

Reply:
My main concern here is that if we use the wrong operation it will only confuse people in the future :see_no_evil: So, even if it does not directly matter I'd put a value here that at least makes sense from the mathematical point of view.

Reply:
```suggestion
        ResourceFacet facet = new ResourceFacet(region.getRegion(), border, gridSize);
```